openjdk-benchmarks/src/jmh/java/org/conscrypt/benchmarks/ClientSocketBenchmark.java

```diff
@@ -0,0 +1,186 @@
+            @Override
+            SSLSocket newSslSocket(String host, int port) {
+                try {
+                    return (SSLSocket) socketFactory.createSocket(
+                            SocketFactory.getDefault().createSocket(host, port), host, port, true);
+                } catch (IOException e) {
+                    throw new RuntimeException(e);
+                }
+            }
+        };
+
+        final SSLSocket newSslSocket(String host, int port, String cipher) {
+            try {
+                SSLSocket sslSocket = newSslSocket(host, port);
+                sslSocket.setEnabledProtocols(getProtocols());
+                sslSocket.setEnabledCipherSuites(new String[] {cipher});
+                return sslSocket;
+            } catch (Exception e) {
+                throw new RuntimeException(e);
+            }
+        }
+
+        abstract SSLSocket newSslSocket(String host, int port);
+    }
+
+    /**
+     * Various factories for the raw socket that backs the SSL socket.
+     */
+    public enum SocketType {
+        DEFAULT {
+            @Override
+            Socket newSocket(String host, int port) {
+                try {
+                    return SocketFactory.getDefault().createSocket(host, port);
+                } catch (IOException e) {
+                    throw new RuntimeException(e);
+                }
+            }
+        },
+        CHANNEL {
+            @Override
+            Socket newSocket(String host, int port) {
+                try {
+                    SocketChannel socketChannel =
+                            SocketChannel.open(new InetSocketAddress(host, port));
+                    return socketChannel.socket();
+                } catch (IOException e) {
+                    throw new RuntimeException(e);
+                }
+            }
+        };
+
+        abstract Socket newSocket(String host, int port);
+    }
+
+    @Param public SslSocketType sslSocketType;
+
+    @Param({"64", "128", "512", "1024", "4096"}) public int messageSize;
+
+    @Param({"TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256"}) public String cipher;
+
+    @Param SocketType socketType;
+
+    private TestClient client;
+    private NettyEchoServer server;
+    private byte[] message;
+
+    @Setup
+    public void setup() throws Exception {
+        message = newTextMessage(messageSize);
+
+        int port = pickUnusedPort();
+        server = new NettyEchoServer(port, messageSize, cipher);
+        server.start();
+
+        client = new TestClient(sslSocketType.newSslSocket(LOCALHOST, port, cipher), messageSize);
+        client.start();
+    }
+
+    @TearDown
+    public void teardown() throws Exception {
+        client.stop();
+        server.stop();
+    }
+
+    @Benchmark
+    public void pingPong() throws IOException {
+        client.sendMessage(message);
+        byte[] output = client.readMessage();
+        assertTrue(Arrays.equals(message, output));
```

Comment:
If `.readMessage` returns an int, you can change this to:

```java
int bytesRead = client.readMessage(recvBuffer);
assertEquals(messageSize, bytesRead);
```

This might also need a `BlackHole` to consume the output, but we can do that later.